File: src/FourHUnfolder.Geometry/Algorithms/PieceComputer.cs
using FourHUnfolder.Domain.Models;

namespace FourHUnfolder.Geometry.Algorithms;

/// <summary>
/// Computes connected components of the fold graph.
/// Each component = one independent paper piece.
/// The component key is the smallest face ID in the group.
/// </summary>
public class PieceComputer
{
    public List<List<int>> ComputePieces(Mesh mesh)
    {
        var uf = new UnionFind(mesh.Faces.Count);

        foreach (var edge in mesh.Edges)
        {
            if (edge.Type == EdgeType.Fold && edge.ConnectsFaces)
                uf.Union(edge.FaceA, edge.FaceB);
        }

        return mesh.Faces
            .GroupBy(f => uf.Find(f.Id))
            .OrderBy(g => g.Key)
            .Select(g => g.Select(f => f.Id).ToList())
            .ToList();
    }
}

File: 4h-unfolder-mac-swift/Sources/FourHUnfolderCore/Core/Algorithms/PieceComputer.swift
// MARK: - Identifies connected paper pieces using Union-Find over fold edges
//
// Mirrors C# PieceComputer.

struct PieceComputer {
    func computePieces(mesh: Mesh) -> [[Int]] {
        let uf = UnionFind(count: mesh.faces.count)

        for edge in mesh.edges where edge.type == .fold && edge.connectsFaces {
            uf.union(edge.faceA, edge.faceB)
        }

        // Group face IDs by their component root
        var groups: [Int: [Int]] = [:]
        for face in mesh.faces {
            let root = uf.find(face.id)
            groups[root, default: []].append(face.id)
        }

        // Sort groups by root, sort face IDs within each group
        return groups
            .sorted { $0.key < $1.key }
            .map { $0.value.sorted() }
    }
}
